### ISSUE TITLE
Printing some C++ code does not syntax highlight.

### ISSUE THREAD

**Issue by dctian** [PDConSec/vsc-print#179]
Filename: foo.cc

This C++ code would print with syntax highlight.  
```
std::vector<std::string> foo() {
};
```

This code would NOT print with syntax highlight.
```
std::map<std::string, std::string> foo() {
};
```

Removing the 2nd argument from `std::map` would make it syntax highlight again.
```
std::map<std::string> foo() {
};
```

I am using version v0.10.20.

**Comment by PeterWone** [comment]
I updated Print to use the latest version of `highlightjs` but this does not entirely resolve the mishandling of multi-type generic syntax.

**Comment by dctian** [comment]
Thanks for the pointer.  I have filed an issue at highlight.js here.  https://github.com/highlightjs/highlight.js/issues/3696#issue-1551726916  Please double check my write up to ensure I got all the details correctly.

It is pretty surprising that basic data structure like std::map<std::string, std::string> would not highlight correctly for highlight.js...

BTW, Are you using `highlight` or `highlightauto`?   this is one of the questions asked me when I filed the issue.

**Comment by PeterWone** [comment]
CPP. 

I know why you asked - their template wants it. VS Code language identifiers are mostly the same as for highlightjs, so Print tries the VS Code language identifier and checks the response. When the response doesn't contain any CSS classes starting with `hljs-` Print falls back to highlightauto.

To give you a definite answer and because we both know the first thing they'll say is why aren't you using the latest version, I bumped the version up and stepped through in the debugger. Enough syntax highlighting is done that Print doesn't try Auto so the answer is a definite CPP.

**Comment by dctian** [comment]
https://github.com/highlightjs/highlight.js/issues/3696  can you update this issue, so that you can correct any misstatement made by me?  thanks so much!

**Comment by dctian** [comment]
https://github.com/highlightjs/highlight.js/issues/3696#issuecomment-1399539102

Hi Peter, given the latest comment in highlight.js, does it seem like its latest branch has the correct rendering logic for C++?  let me know.

**Comment by PeterWone** [comment]
He said "latest branch" which may be newer than their latest release. I'll have to get their latest branch and have a look. At the moment I'm at work (in Australia).
________________________________

**Comment by PeterWone** [comment]
I just checked this again and the problem is not fully resolved. 
![image](https://user-images.githubusercontent.com/5498936/220586800-e29e9d57-de69-4ae6-b50c-2becf33840fa.png)
![image](https://user-images.githubusercontent.com/5498936/220586945-6a6fc954-5532-44d0-ba2c-49e471d80d89.png)

**Comment by PeterWone** [comment]
David, as you probably know from the related issue on highlightjs repo, they regard this as fixed in v11 but I can't use v11 for assorted reasons. In the short term this means that I can't do anything about this issue. In the long term it seems like I'll have to find an alternative way to render.

I'm going to close this issue because there isn't any short-term resolution that I can pursue.

If it becomes possible to use v11, I will. Otherwise, I'll try to find an alternative.

**Comment by dctian** [comment]
I wonder what is the syntax highlight engine used by VS code itself?  Can you tap into that since it works very well for all languages and should be supported by Microsoft and VS code maintainers.

Just a thought. Thanks.

David

On Mar 5, 2023, at 11:33 PM, Peter Wone ***@***.***> wrote:

﻿

Closed #179<https://github.com/PDConSec/vsc-print/issues/179> as completed.

**Comment by PeterWone** [comment]
That's an astute question with a long and complicated answer. It's probably the _right_ answer in the long term but it's not trivial to use. The way VS Code does syntax colouring is event driven and processes only the portion of the code that is actually being displayed.
